pkg/controllers/subnetbinding/controller.go

```diff
@@ -0,0 +1,353 @@
+func (r *Reconciler) validateDependency(ctx context.Context, bindingMap *v1alpha1.SubnetConnectionBindingMap) (*model.VpcSubnet, []*model.VpcSubnet, *errorWithRetry) {
+	childSubnets, err := r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.SubnetName, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	childSubnet := childSubnets[0]
+
+	if bindingMap.Spec.TargetSubnetName != "" {
+		parentSubnets, err := r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetName, true)
+		if err != nil {
+			return nil, nil, err
+		}
+		return childSubnet, parentSubnets, nil
+	}
+
+	parentSubnets, err := r.validateVpcSubnetsBySubnetSetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetSetName)
+	if err != nil {
+		return nil, nil, err
+	}
+	return childSubnet, parentSubnets, nil
+}
+
+func (r *Reconciler) validateVpcSubnetsBySubnetCR(ctx context.Context, namespace, name string, isTarget bool) ([]*model.VpcSubnet, *errorWithRetry) {
+	subnetCR := &v1alpha1.Subnet{}
+	subnetKey := types.NamespacedName{Namespace: namespace, Name: name}
+	// Check the Subnet CR existence.
+	err := r.Client.Get(ctx, subnetKey, subnetCR)
+	if err != nil {
+		log.Error(err, "Failed to get Subnet CR", "key", subnetKey.String())
+		return nil, &errorWithRetry{
+			message: fmt.Sprintf("Unable to get Subnet CR %s", name),
+			retry:   false,
+			error:   fmt.Errorf("failed to get subnet %s in Namespace %s with error: %v", name, namespace, err),
+		}
+	}
+
+	// Check the Subnet CR realization.
+	subnets := r.SubnetService.ListSubnetCreatedBySubnet(string(subnetCR.UID))
+	if len(subnets) == 0 {
+		log.Info("NSX VpcSubnets by subnet CR do not exist", "key", subnetKey.String())
+		return nil, &errorWithRetry{
+			message: fmt.Sprintf("Subnet CR %s is not realized on NSX", name),
+			retry:   false,
+			error:   fmt.Errorf("not found NSX VpcSubnets created by Subnet CR '%s/%s'", namespace, name),
+		}
+	}
+
+	// Check if the Subnet CR is nested.
+	if !isTarget {
+		bms := r.SubnetBindingService.GetSubnetConnectionBindingMapsByParentSubnet(subnets[0])
+		if len(bms) > 0 {
+			dependency := r.SubnetBindingService.GetCRNameBySubnetConnectionBindingMap(bms[0])
+			return nil, &errorWithRetry{
+				message: fmt.Sprintf("Subnet CR %s is working as target by %s", name, dependency),
+				error:   fmt.Errorf("Subnet %s already works as target in SegmentConnectionBindingMap %s", name, dependency),
+				retry:   true,
+			}
+		}
+	} else {
+		bms := r.SubnetBindingService.GetSubnetConnectionBindingMapsByChildSubnet(subnets[0])
+		if len(bms) > 0 {
+			dependency := r.SubnetBindingService.GetCRNameBySubnetConnectionBindingMap(bms[0])
+			return nil, &errorWithRetry{
+				message: fmt.Sprintf("Target Subnet CR %s is attached by %s", name, dependency),
+				error:   fmt.Errorf("target Subnet %s is already attached by SegmentConnectionBindingMap %s", name, dependency),
```

Comment:
Why is this error not empty if Subnet is allowed to attached by multiple SubnetConnectionBindingMap?